src/main/java/io/supertokens/cronjobs/bulkimport/ProcessBulkImportUsers.java

```diff
@@ -0,0 +1,597 @@
+            throws TenantOrAppNotFoundException, InvalidConfigException, IOException, DbInitException, StorageQueryException {
+        List<Storage> allProxyStorages = new ArrayList<>();
+
+        Map<ResourceDistributor.KeyClass, ResourceDistributor.SingletonResource> resources = main
+                .getResourceDistributor()
+                .getAllResourcesWithResourceKey(StorageLayer.RESOURCE_KEY);
+        for (ResourceDistributor.KeyClass key : resources.keySet()) {
+            if (key.getTenantIdentifier().toAppIdentifier().equals(appIdentifier)) {
+                allProxyStorages.add(getProxyStorage(key.getTenantIdentifier()));
+            }
+        }
+        return allProxyStorages.toArray(new Storage[0]);
+    }
+
+    private void closeAllProxyStorages() throws StorageQueryException {
+        for (SQLStorage storage : userPoolToStorageMap.values()) {
+            storage.closeConnectionForBulkImportProxyStorage();
+            storage.close();
+        }
+        userPoolToStorageMap.clear();
+    }
+
+    private void processUser(AppIdentifier appIdentifier, BulkImportUser user, BulkImportUserUtils bulkImportUserUtils,
+            BulkImportSQLStorage baseTenantStorage)
+            throws TenantOrAppNotFoundException, StorageQueryException, InvalidConfigException, IOException,
+            DbInitException {
+
+        try {
+            if (Main.isTesting && Main.isTesting_skipBulkImportUserValidationInCronJob) {
+                // Skip validation when the flag is enabled during testing
+            } else {
+                // Validate the user
+                bulkImportUserUtils.createBulkImportUserFromJSON(main, appIdentifier, user.toJsonObject(), user.id);
+            }
+
+            // Since all the tenants of a user must share the storage, we will just use the
+            // storage of the first tenantId of the first loginMethod
+
+            TenantIdentifier firstTenantIdentifier = new TenantIdentifier(appIdentifier.getConnectionUriDomain(),
+                    appIdentifier.getAppId(), user.loginMethods.get(0).tenantIds.get(0));
+
+            SQLStorage bulkImportProxyStorage = (SQLStorage) getProxyStorage(firstTenantIdentifier);
+
+            LoginMethod primaryLM = getPrimaryLoginMethod(user);
+
+            AuthRecipeSQLStorage authRecipeSQLStorage = (AuthRecipeSQLStorage) getProxyStorage(firstTenantIdentifier);
+            // If primaryUserId is not null, it means we may have already processed this user but failed to delete the entry
+            // If the primaryUserId exists in the database, we'll delete the corresponding entry from the bulkImportUser table and proceed to skip this user.
+            if (user.primaryUserId != null) {
+                AuthRecipeUserInfo processedUser = authRecipeSQLStorage.getPrimaryUserById(appIdentifier,
+                        user.primaryUserId);
+
+                if (processedUser != null && isProcessedUserFromSameBulkImportUserEntry(processedUser, user)) {
```

Comment:
isProcessedUserFromSameBulkImportUserEntry need way more commenting / explanation. In fact, the whole way in which the code is trying to ensure consistency needs to be reexplained with an example for different cases. As a reader of the code, it's really hard to understand what's happening and why

Reply:
Added a comment to explain the whole flow.